### PR TITLE
Correct handling of string functions for Unicode (non-ASCII) strings

### PR DESCRIPTION
**Pull request by dain**
Currently, all of the string functions work incorrectly for Unicode (non-ASCII) strings (see  #2579)
To fix this, @fmeiser rewrote most of the methods to work correctly for Unicode. For the methods which already worked with non-ASCII, he added tests. All methods work directly with `Slice` without converting to `String`.

I split original pull request (#2591) into two parts.  One part, I added to Slice directly (https://github.com/airlift/slice/pull/38), so it could access the internal of Slice, which result in a small speed boost.  The other part is here.  Additionally, I added more tests and a benchmark.

Benchmark results
```
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                          true         2  avgt        5      20.222       18.583  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                          true         5  avgt        5      31.776       14.151  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                          true        10  avgt        5      59.151       25.197  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                          true       100  avgt        5     445.298       44.887  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                          true      1000  avgt        5    4220.048      387.372  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                          true     10000  avgt        5   41436.997     6748.643  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                         false         2  avgt        5      36.761       19.102  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                         false         5  avgt        5      60.048       15.565  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                         false        10  avgt        5     100.922       24.394  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                         false       100  avgt        5     733.359       22.232  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                         false      1000  avgt        5    8330.696      177.062  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLeftTrim                         false     10000  avgt        5  105107.335     5595.990  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                            true         2  avgt        5       4.927        0.153  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                            true         5  avgt        5       4.568        0.140  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                            true        10  avgt        5       6.630        0.250  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                            true       100  avgt        5      33.386        0.911  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                            true      1000  avgt        5     224.277       10.458  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                            true     10000  avgt        5    2205.505      161.019  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                           false         2  avgt        5       4.080        0.297  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                           false         5  avgt        5      13.217        0.323  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                           false        10  avgt        5      14.578        0.711  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                           false       100  avgt        5      80.661        2.248  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                           false      1000  avgt        5     761.991       30.133  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLength                           false     10000  avgt        5    8226.699       81.321  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                             true         2  avgt        5      42.576        2.118  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                             true         5  avgt        5      65.601        3.301  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                             true        10  avgt        5     102.730       11.801  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                             true       100  avgt        5     769.916       17.984  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                             true      1000  avgt        5    8224.269      217.953  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                             true     10000  avgt        5   74965.387     4825.826  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                            false         2  avgt        5      52.088        0.915  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                            false         5  avgt        5     105.048        8.148  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                            false        10  avgt        5     193.123       13.632  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                            false       100  avgt        5    1569.459      127.311  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                            false      1000  avgt        5   17690.420     1098.144  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkLower                            false     10000  avgt        5  185126.292    10332.891  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                           true         2  avgt        5      25.245        9.516  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                           true         5  avgt        5      38.720       21.166  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                           true        10  avgt        5      57.678       30.383  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                           true       100  avgt        5     343.552        5.923  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                           true      1000  avgt        5    3676.776      540.327  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                           true     10000  avgt        5   46689.527    13522.092  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                          false         2  avgt        5      28.990       16.642  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                          false         5  avgt        5      46.178       27.107  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                          false        10  avgt        5      79.428       80.178  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                          false       100  avgt        5     508.880      298.995  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                          false      1000  avgt        5    5623.366     1715.816  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkReverse                          false     10000  avgt        5   75638.802    19926.038  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                         true         2  avgt        5      13.378        5.616  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                         true         5  avgt        5      22.950       19.878  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                         true        10  avgt        5      31.974       25.273  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                         true       100  avgt        5     212.902       12.214  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                         true      1000  avgt        5    1888.830       94.442  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                         true     10000  avgt        5   18814.020     1200.260  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                        false         2  avgt        5      36.954       20.222  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                        false         5  avgt        5      70.683       16.365  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                        false        10  avgt        5     122.992       15.054  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                        false       100  avgt        5     721.577      144.570  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                        false      1000  avgt        5    7427.438      583.645  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkRightTrim                        false     10000  avgt        5   98015.484     6136.921  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                    true         2  avgt        5       2.798        0.163  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                    true         5  avgt        5       5.267        0.144  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                    true        10  avgt        5      18.864       11.269  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                    true       100  avgt        5      29.027       19.471  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                    true      1000  avgt        5     157.819       27.130  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                    true     10000  avgt        5    1280.307      100.259  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                   false         2  avgt        5       2.803        0.090  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                   false         5  avgt        5       5.269        0.243  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                   false        10  avgt        5      33.788       16.033  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                   false       100  avgt        5      94.022       16.719  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                   false      1000  avgt        5     619.170       44.939  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStart                   false     10000  avgt        5    5476.401      266.483  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd             true         2  avgt        5       8.878        0.519  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd             true         5  avgt        5      20.183        7.732  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd             true        10  avgt        5      21.114        5.898  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd             true       100  avgt        5      73.295       28.279  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd             true      1000  avgt        5     383.032       26.316  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd             true     10000  avgt        5    3507.865      183.545  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd            false         2  avgt        5       7.596        0.398  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd            false         5  avgt        5      39.385       28.400  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd            false        10  avgt        5      52.014       22.878  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd            false       100  avgt        5     177.969       25.920  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd            false      1000  avgt        5    1325.836      137.824  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartFromEnd            false     10000  avgt        5   12168.301     1280.519  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength              true         2  avgt        5       2.779        0.179  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength              true         5  avgt        5      26.832       12.733  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength              true        10  avgt        5      32.499       22.123  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength              true       100  avgt        5      51.984        9.618  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength              true      1000  avgt        5     274.754       21.501  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength              true     10000  avgt        5    2513.272      160.565  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength             false         2  avgt        5       2.808        0.121  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength             false         5  avgt        5      47.729       23.736  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength             false        10  avgt        5      65.283       23.358  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength             false       100  avgt        5     160.441       17.009  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength             false      1000  avgt        5     965.220       58.410  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLength             false     10000  avgt        5    9237.194      497.736  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd       true         2  avgt        5      26.134       10.500  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd       true         5  avgt        5      36.046       17.350  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd       true        10  avgt        5      38.347       14.512  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd       true       100  avgt        5      90.389       25.483  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd       true      1000  avgt        5     511.703       27.230  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd       true     10000  avgt        5    4750.136      440.145  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd      false         2  avgt        5      35.049       20.179  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd      false         5  avgt        5      67.110       24.906  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd      false        10  avgt        5      97.434       22.444  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd      false       100  avgt        5     267.049       27.025  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd      false      1000  avgt        5    1842.101      140.554  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkSubstringStartLengthFromEnd      false     10000  avgt        5   17421.955      525.860  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                              true         2  avgt        5      14.986        0.743  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                              true         5  avgt        5      27.355        0.339  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                              true        10  avgt        5      44.679        1.630  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                              true       100  avgt        5     415.864       23.629  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                              true      1000  avgt        5    3735.030      163.791  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                              true     10000  avgt        5   42947.000    21308.508  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                             false         2  avgt        5      23.956        1.923  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                             false         5  avgt        5      41.364        3.133  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                             false        10  avgt        5      69.545        3.745  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                             false       100  avgt        5     609.266       16.541  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                             false      1000  avgt        5    7155.778      334.445  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkTrim                             false     10000  avgt        5  101488.696     5323.441  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                             true         2  avgt        5      46.959       37.490  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                             true         5  avgt        5      67.280       52.553  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                             true        10  avgt        5     103.150       58.757  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                             true       100  avgt        5     704.528       38.133  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                             true      1000  avgt        5    7471.537      437.504  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                             true     10000  avgt        5   69990.883     4131.438  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                            false         2  avgt        5      62.522        2.174  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                            false         5  avgt        5      93.060       56.668  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                            false        10  avgt        5     148.329       69.036  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                            false       100  avgt        5    1662.402       84.428  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                            false      1000  avgt        5   17539.205      495.578  ns/op
c.f.p.o.s.StringFunctionsBenchmark.benchmarkUpper                            false     10000  avgt        5  186298.605    10171.332  ns/op
```

### CODE REVIEW

**Comment by dain** [comment]
@fmeiser can you review this and let me know if I messed anything up?

File: presto-docs/src/main/sphinx/functions/string.rst [hunk 1]
```diff
@@ -36,9 +40,15 @@ String Functions
 
     Converts ``string`` to lowercase.
 
+.. warning::
+
+    This method does not perform perform locale-sensitive, context-sensitive,
```
Comment (haozhun) [inline presto-docs/src/main/sphinx/functions/string.rst]: typo

File: presto-docs/src/main/sphinx/functions/string.rst [hunk 2]
```diff
@@ -36,9 +40,15 @@ String Functions
...
+    or one-to-many mappings required for some languages.  Specifically, this
+    will return incorrect results for Lithuanian, Turkish, and Azeri.
+
 .. function:: ltrim(string) -> varchar
 
-    Removes leading spaces from ``string``.
+    Removes leading white space from ``string``.
```
Comment (haozhun) [inline presto-docs/src/main/sphinx/functions/string.rst]: white spaces

File: presto-docs/src/main/sphinx/release/release-0.101.rst [hunk 3]
```diff
@@ -2,6 +2,15 @@
...
 =============
 
+Unicode Support
+---------------
+
+All varchar functions have been updated to support Unicode. The functions assume
+that the varchar contains valid UTF-8 encoded code points. There are no explicit
+checks for valid UTF-8, and the function may return incorrect results or throw an
+fail on invalid UTF-8.  Invalid UTF-8 data can be corrected with the :func:`from_utf8`
+function.
```
Comment (haozhun) [inline presto-docs/src/main/sphinx/release/release-0.101.rst]: Maybe mention grapheme clusters here?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 4]
```diff
@@ -91,22 +93,78 @@ public static Slice replace(@SqlType(StandardTypes.VARCHAR) Slice str, @SqlType(
...
     public static Slice replace(@SqlType(StandardTypes.VARCHAR) Slice str, @SqlType(StandardTypes.VARCHAR) Slice search, @SqlType(StandardTypes.VARCHAR) Slice replace)
     {
-        String replaced = str.toString(UTF_8).replace(
-                search.toString(UTF_8),
-                replace.toString(UTF_8));
-        return Slices.copiedBuffer(replaced, UTF_8);
+        // Empty search?
+        if (search.length() == 0) {
+            // With empty `search` we insert `replace` in front of every character and and the end
+            Slice buffer = Slices.allocate((countCodePoints(str) + 1) * replace.length() + str.length());
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Should use str.length() instead.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 5]
```diff
@@ -252,58 +395,82 @@ public static Slice rightTrim(@SqlType(StandardTypes.VARCHAR) Slice slice)
...
+    @Description("decodes the UTF-8 encoded string")
     @ScalarFunction
     @SqlType(StandardTypes.VARCHAR)
-    public static Slice upper(@SqlType(StandardTypes.VARCHAR) Slice slice)
+    public static Slice fromUtf8(@SqlType(StandardTypes.VARBINARY) Slice slice, @SqlType(StandardTypes.BIGINT) long replacementCodePoint)
     {
-        Slice upper = Slices.allocate(slice.length());
-        for (int i = 0; i < slice.length(); i++) {
-            upper.setByte(i, Ascii.toUpperCase((char) slice.getByte(i)));
+        if (replacementCodePoint <= 0x10FFFF) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: This line seems weird to me. What is this code trying to do?

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/StringFunctionsBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        static {
+            ASCII_CODE_POINTS = new int[0x7F];
+            for (int i = 0; i < ASCII_CODE_POINTS.length; i++) {
+                ASCII_CODE_POINTS[i] = i;
+            }
+
+            int[] codePoints = new int[MAX_CODE_POINT];
+            int index = 0;
+            for (int codePoint = 0; codePoint < MAX_CODE_POINT; codePoint++) {
+                if (Character.isDefined(codePoint)) {
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/StringFunctionsBenchmark.java]: Looks like this code is not up to date. As is return, surrogate characters will be included.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/StringFunctionsBenchmark.java [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        @Setup
+        public void setup()
+        {
+            Slice whitespace = createRandomUtf8Slice(ascii ? ASCII_WHITESPACE : ALL_WHITESPACE, length + 1);
+            leftWhitespace = Slices.copyOf(whitespace);
+            leftWhitespace.setByte(leftWhitespace.length() - 1, 'X');
+            rightWhitespace = Slices.copyOf(whitespace);
+            rightWhitespace.setByte(0, 'X');
+            bothWhitespace = Slices.copyOf(whitespace);
+            rightWhitespace.setByte(length / 2, 'X');
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/StringFunctionsBenchmark.java]: typo

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java [hunk 8]
```diff
@@ -121,16 +164,41 @@ public void testSubstring()
...
+        assertFunction("SPLIT('a.b.c.', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "c."));
+        // Test SPLIT for non-ASCII
+        assertFunction("SPLIT('\u4FE1\u5FF5,\u7231,\u5E0C\u671B', ',', 3)", new ArrayType(VARCHAR), ImmutableList.of("\u4FE1\u5FF5", "\u7231", "\u5E0C\u671B"));
+        assertFunction("SPLIT('\u8B49\u8BC1\u8A3C', '\u8BC1', 2)", new ArrayType(VARCHAR), ImmutableList.of("\u8B49", "\u8A3C"));
+
+        assertFunction("SPLIT('.a.b.c', '.', 4)", new ArrayType(VARCHAR), ImmutableList.of("", "a", "b", "c"));
+        assertFunction("SPLIT('.a.b.c', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("", "a", "b.c"));
+        assertFunction("SPLIT('.a.b.c', '.', 2)", new ArrayType(VARCHAR), ImmutableList.of("", "a.b.c"));
+        assertFunction("SPLIT('a..b..c', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("a", "", "b..c"));
+        assertFunction("SPLIT('a.b..', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "."));
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: * Add test for `SPLIT('..a...a..', '.')`. My regexp split will produce `["", "", "a", "", "", "a", "", ""]`.
* Also test for `SPLIT('...', '.')`. My regexp split will produce `["", "", "", ""]`.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java [hunk 9]
```diff
@@ -189,6 +268,10 @@ public void testLeftTrim()
         assertFunction("LTRIM('  hello')", VARCHAR, "hello");
         assertFunction("LTRIM('hello  ')", VARCHAR, "hello  ");
         assertFunction("LTRIM(' hello world ')", VARCHAR, "hello world ");
+
+        assertFunction("LTRIM('\u4FE1\u5FF5 \u7231 \u5E0C\u671B  ')", VARCHAR, "\u4FE1\u5FF5 \u7231 \u5E0C\u671B  ");
+        assertFunction("LTRIM(' \u4FE1\u5FF5 \u7231 \u5E0C\u671B ')", VARCHAR, "\u4FE1\u5FF5 \u7231 \u5E0C\u671B ");
+        assertFunction("LTRIM('  \u4FE1\u5FF5 \u7231 \u5E0C\u671B')", VARCHAR, "\u4FE1\u5FF5 \u7231 \u5E0C\u671B");
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: Add tests for unicode spaces, and ascii non-32 spaces. Same for `RTRIM`, `TRIM`.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java [hunk 10]
```diff
@@ -227,5 +320,28 @@ public void testUpper()
         assertFunction("UPPER('')", VARCHAR, "");
         assertFunction("UPPER('Hello World')", VARCHAR, "HELLO WORLD");
         assertFunction("UPPER('what!!')", VARCHAR, "WHAT!!");
+        assertFunction("UPPER('\u00D6sterreich')", VARCHAR, upperByCodePoint("\u00D6STERREICH"));
+        assertFunction("UPPER('From\uD801\uDC2DTo')", VARCHAR, upperByCodePoint("FROM\uD801\uDC2DTO"));
+    }
+
+    private static String lowerByCodePoint(String string)
+    {
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: Add a short comment on why `String.toLowerCase` is different.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java [hunk 11]
```diff
@@ -227,5 +320,28 @@ public void testUpper()
...
+        return new String(upperCodePoints, 0, upperCodePoints.length);
+    }
+
+    @Test
+    public void testFromUtf8()
+    {
+        assertFunction("from_utf8(to_utf8('hello'))", VARCHAR, "hello");
+        assertFunction("from_utf8(from_hex('58BF'))", VARCHAR, "X\uFFFD");
+        assertFunction("from_utf8(from_hex('58DF'))", VARCHAR, "X\uFFFD");
+        assertFunction("from_utf8(from_hex('58F7'))", VARCHAR, "X\uFFFD");
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: Add a test of from_utf8 using custom replacement character.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java (hunk 8 above)
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: @electrum do you agree with this?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 4 above)
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Are you sure... this code is `(replaceLength * (numberOfCodePointsInStr + 1)) + strLength)`  so basically we will end up with the full string plus one replacement per code point, with one extra at the end.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java (hunk 8 above)
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: The function currently does not allow for a trailing empty part.

Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: After discussion with @martint, the behavior of regexp split was changed to allow trailing empty part. I'm not sure if it's a good idea to keep the behavior of this split and split of regexp the same.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 4 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: I meant `Slices.allocate((str.length() + 1) * replace.length() + str.length())`. Number of code points is irrelevant here.
